src/application.c

```diff
@@ -45,6 +45,36 @@ static void update_ui_panel_toggle_button(struct swappy_state *state) {
   gtk_widget_set_visible(painting_box, toggled);
 }
 
+void application_finish(struct swappy_state *state) {
+  paint_free_all(state);
+  pixbuf_free(state);
+  cairo_surface_destroy(state->rendering_surface);
+  cairo_surface_destroy(state->original_image_surface);
+  if (state->temp_file_str) {
+    g_info("deleting temporary file: %s", state->temp_file_str);
+    if (g_unlink(state->temp_file_str) != 0) {
+      g_warning("unable to delete temporary file: %s", state->temp_file_str);
+    }
+    g_free(state->temp_file_str);
+  }
+  g_free(state->file_str);
+  g_free(state->geometry);
+  g_free(state->window);
+  g_free(state->ui);
+
+  g_object_unref(state->app);
+
+  config_free(state);
+}
+
+static void early_exit(struct swappy_state *state){
```

Comment:
You don't need this function. It's actually a bit misleading from the function calling this because it doesn't always exit the app.

Reply:
Calling `gtk_main_quit` will exit the main loop and call `application_finish` in `main.c`